File: AMAppkit/UITableView+Reloading.m
//
//  UITableView+Reloading.m
//  AMAppkit
//
//  Created by Ilya Kuznecov on 26/01/2017.
//  Copyright © 2017 Arello Mobile. All rights reserved.
//

#import "UITableView+Reloading.h"

@implementation UITableView (Reloading)

- (void)reloadWithOldData:(NSArray*)oldData newData:(NSArray*)newData block:(dispatch_block_t)block addAnimation:(UITableViewRowAnimation)animation {
    NSMutableArray *deleteIndexPaths = [NSMutableArray array];
    
    NSSet *oldDataSet = [NSSet setWithArray:oldData];
    NSSet *newDataSet = [NSSet setWithArray:newData];
    
    for (NSUInteger index = 0; index < oldData.count; index++) {
        if (![newDataSet containsObject:oldData[index]]) {
            [deleteIndexPaths addObject:[NSIndexPath indexPathForRow:index inSection:0]];
        }
    }
    NSMutableArray *addIndexPaths = [NSMutableArray array];
    for (NSUInteger index = 0; index < newData.count; index++) {
        if (![oldDataSet containsObject:newData[index]]) {
            [addIndexPaths addObject:[NSIndexPath indexPathForRow:index inSection:0]];
        }
    }
    
    [self beginUpdates];
    
    if (deleteIndexPaths.count) {
        [self deleteRowsAtIndexPaths:deleteIndexPaths withRowAnimation:UITableViewRowAnimationFade];
    }
    if (addIndexPaths.count) {
        [self insertRowsAtIndexPaths:addIndexPaths withRowAnimation:animation];
    }
    
    if (block) {
        block();
    }
    [self endUpdates];
}

@end

File: AMAppkit/UITableView+Reloading.swift
//
//  UITableView+Reloading.swift
//  AMAppkit
//
//  Created by Ilya Kuznetsov on 6/1/18.
//

import UIKit

public extension UITableView {
    
    private func printDuplicates(_ array: [AnyHashable]) {
        var allSet = Set<AnyHashable>()
        
        for object in array {
            if allSet.contains(object) {
                print("found duplicated object %@", object.description)
            } else {
                allSet.insert(object)
            }
        }
    }
    
    func reload(oldData: [AnyHashable], newData: [AnyHashable], deferred: (()->())?, addAnimation: UITableView.RowAnimation) {
        
        var toAdd: [IndexPath] = []
        var toDelete: [IndexPath] = []
        
        let oldDataSet = Set(oldData)
        let newDataSet = Set(newData)
        
        if oldDataSet.count != oldData.count {
            printDuplicates(oldData)
        }
        if newDataSet.count != newData.count {
            printDuplicates(newData)
        }
        
        let currentSet = NSMutableOrderedSet(array: oldData)
        for (index, object) in oldData.enumerated() {
            if !newDataSet.contains(object) {
                toDelete.append(IndexPath(row: index, section: 0))
                currentSet.remove(object)
            }
        }
        for (index, object) in newData.enumerated() {
            if !oldDataSet.contains(object) {
                toAdd.append(IndexPath(row: index, section: 0))
                currentSet.insert(object, at: index)
            }
        }
        
        var itemsToMove: [(from: IndexPath, to: IndexPath)] = []
        for (index, object) in newData.enumerated() {
            let oldDataIndex = currentSet.index(of: object)
            if index != oldDataIndex {
                itemsToMove.append((from: IndexPath(row: oldData.index(of: object)!, section: 0), to: IndexPath(row: index, section: 0)))
            }
        }
        
        self.beginUpdates()
        
        if !toDelete.isEmpty {
            self.deleteRows(at: toDelete, with: .fade)
        }
        if !toAdd.isEmpty {
            self.insertRows(at: toAdd, with: addAnimation)
        }
        if !itemsToMove.isEmpty {
            for couple in itemsToMove {
                self.moveRow(at: couple.from, to: couple.to)
            }
        }
        deferred?()
        
        self.endUpdates()
    }
}
